File: dp/c/domain_event.c
// Domain Event Pattern in C

#include <stdio.h>
#include <string.h>

typedef enum { EVT_ORDER_PLACED, EVT_ORDER_SHIPPED, EVT_PAYMENT_RECEIVED } DomainEventType;

typedef struct {
    DomainEventType type;
    int order_id;
    char data[64];
    double amount;
} DomainEvent;

typedef void (*EventHandler)(const DomainEvent*);

#define MAX_HANDLERS 16
typedef struct {
    EventHandler handlers[MAX_HANDLERS];
    int count;
} EventBus;

void bus_init(EventBus* b) { b->count = 0; }
void bus_subscribe(EventBus* b, EventHandler h) { b->handlers[b->count++] = h; }

void bus_publish(EventBus* b, const DomainEvent* e) {
    for (int i = 0; i < b->count; i++)
        b->handlers[i](e);
}

void email_handler(const DomainEvent* e) {
    if (e->type == EVT_ORDER_PLACED)
        printf("  [Email] Confirmation for order #%d\n", e->order_id);
}

void accounting_handler(const DomainEvent* e) {
    if (e->type == EVT_PAYMENT_RECEIVED)
        printf("  [Accounting] $%.2f for order #%d\n", e->amount, e->order_id);
}

void audit_handler(const DomainEvent* e) {
    const char* names[] = {"OrderPlaced", "OrderShipped", "PaymentReceived"};
    printf("  [Audit] %s #%d\n", names[e->type], e->order_id);
}

int main(void) {
    EventBus bus;
    bus_init(&bus);
    bus_subscribe(&bus, email_handler);
    bus_subscribe(&bus, accounting_handler);
    bus_subscribe(&bus, audit_handler);

    DomainEvent e1 = {EVT_ORDER_PLACED, 1001, "Alice", 99.99};
    printf("Publishing OrderPlaced:\n");
    bus_publish(&bus, &e1);

    DomainEvent e2 = {EVT_PAYMENT_RECEIVED, 1001, "", 99.99};
    printf("\nPublishing PaymentReceived:\n");
    bus_publish(&bus, &e2);
    return 0;
}

File: dp/rust/src/domain_event.rs
// Domain Event Pattern in Rust
// Capture state changes as events for decoupled communication

use std::fmt;

#[derive(Debug, Clone)]
struct Timestamp(String);

impl Timestamp {
    fn now() -> Self {
        Timestamp("2024-01-15T10:30:00Z".to_string())
    }
}

// Domain events
#[derive(Debug, Clone)]
enum DomainEvent {
    OrderPlaced { order_id: u32, customer: String, total: f64, at: Timestamp },
    OrderShipped { order_id: u32, tracking: String, at: Timestamp },
    OrderCancelled { order_id: u32, reason: String, at: Timestamp },
    PaymentReceived { order_id: u32, amount: f64, at: Timestamp },
}

impl fmt::Display for DomainEvent {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            DomainEvent::OrderPlaced { order_id, customer, total, .. } =>
                write!(f, "OrderPlaced(#{}, {}, ${:.2})", order_id, customer, total),
            DomainEvent::OrderShipped { order_id, tracking, .. } =>
                write!(f, "OrderShipped(#{}, {})", order_id, tracking),
            DomainEvent::OrderCancelled { order_id, reason, .. } =>
                write!(f, "OrderCancelled(#{}, {})", order_id, reason),
            DomainEvent::PaymentReceived { order_id, amount, .. } =>
                write!(f, "PaymentReceived(#{}, ${:.2})", order_id, amount),
        }
    }
}

// Event bus
type Handler = Box<dyn Fn(&DomainEvent)>;

struct EventBus {
    handlers: Vec<Handler>,
}

impl EventBus {
    fn new() -> Self {
        EventBus { handlers: Vec::new() }
    }

    fn subscribe(&mut self, handler: Handler) {
        self.handlers.push(handler);
    }

    fn publish(&self, event: &DomainEvent) {
        println!("[EventBus] Publishing: {}", event);
        for handler in &self.handlers {
            handler(event);
        }
    }
}

fn main() {
    let mut bus = EventBus::new();

    // Subscribe handlers
    bus.subscribe(Box::new(|event| {
        if let DomainEvent::OrderPlaced { customer, .. } = event {
            println!("  [Email] Confirmation sent to {}", customer);
        }
    }));

    bus.subscribe(Box::new(|event| {
        if let DomainEvent::PaymentReceived { order_id, amount, .. } = event {
            println!("  [Accounting] Recorded ${:.2} for order #{}", amount, order_id);
        }
    }));

    bus.subscribe(Box::new(|event| {
        println!("  [Audit] Event logged: {}", event);
    }));

    // Publish events
    bus.publish(&DomainEvent::OrderPlaced {
        order_id: 1001, customer: "Alice".into(), total: 99.99, at: Timestamp::now(),
    });
    println!();
    bus.publish(&DomainEvent::PaymentReceived {
        order_id: 1001, amount: 99.99, at: Timestamp::now(),
    });
    println!();
    bus.publish(&DomainEvent::OrderShipped {
        order_id: 1001, tracking: "TRK-12345".into(), at: Timestamp::now(),
    });
}
